Add support for font-variant-caps in WidthIterator (simple path)

vitorroriz opened the pull request:
#### 03b288845585ac6ed3440b092b97bd543da6a5be
<pre>
Add support for font-variant-caps in WidthIterator (simple path)
<a href="https://bugs.webkit.org/show_bug.cgi?id=259736">https://bugs.webkit.org/show_bug.cgi?id=259736</a>
rdar://110431790

Reviewed by Myles C. Maxfield.

Right now, font-variant small-caps can just be handled if we take
the Complex path in text layout, which is implemented by ComplexTextController.
Implementing small-caps handling in WidthIterator allows WebKit to
layout text taking the Simple path and use the WidthIterator when
font-feature or font-variants are used, including font-variant-caps.

The logic introduced for small-caps tries to mimic the one
in ComplexTextController. There, everytime there is change
in Font or in the small-caps state, a new run is created
with the determined font, which can be synthesized or not.
If any character in the run needs to be synthesized, all characters
are synthesized.

Here, for the WidthIterator, we don&apos;t split our runs in new runs for
each variation. Instead, we define a range and commit it. Each range
is represented by a font, that can be synthesized or not, like in
ComplexTextController, and a range glyph buffer. The buffer contains
the glyphs to the characters that compose the range. They are
add to the final glyph buffer each time we commit a range.

We are reusing the logic for determinig if a character should be synthesized
or not. Therefore, &quot;shouldSynthesize()&quot; and &quot;capitalized()&quot; functions
are moved from ComplexTextController to FontCascade.

When we are processing a range and we detect that the current font
can&apos;t be used for represent a glyph and it needs to be synthesized,
we need to synthesize the whole range. Therefore, we reset the iterator
to the index of the last font transition we have registered. For that,
we are adding the capability for the Latin1TextIterator and the
SurrogatePairAwareTextIterator to reset the current index to a given
index.

* LayoutTests/platform/gtk/fast/text/drawBidiText-expected.png:
* LayoutTests/platform/gtk/fast/text/drawBidiText-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/svg/import/text-dom-02-f-manual-expected.png: Added.
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/svg/import/text-dom-02-f-manual-expected.txt: Added.
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/svg/import/text-dom-05-f-manual-expected.png: Added.
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/svg/import/text-dom-05-f-manual-expected.txt: Added.
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/svg/import/text-tselect-02-f-manual-expected.png: Added.
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/svg/import/text-tselect-02-f-manual-expected.txt: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/svg/import/text-dom-02-f-manual-expected.png: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/svg/import/text-dom-02-f-manual-expected.txt: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/svg/import/text-dom-05-f-manual-expected.png: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/svg/import/text-dom-05-f-manual-expected.txt: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/svg/import/text-tselect-02-f-manual-expected.png: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/svg/import/text-tselect-02-f-manual-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/import/fonts-desc-02-t-manual.svg:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/text-dom-02-f-manual-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/text-dom-05-f-manual-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/text-tselect-02-f-manual-expected.txt:
* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/import/fonts-desc-02-t-manual-expected.png:
* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/import/fonts-desc-02-t-manual-expected.txt:
* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/import/fonts-desc-05-t-manual-expected.png:
* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/import/fonts-desc-05-t-manual-expected.txt:
* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/import/text-fonts-203-t-manual-expected.txt:
* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/import/text-fonts-204-t-manual-expected.txt:
* LayoutTests/platform/ios/svg/W3C-SVG-1.1/fonts-desc-02-t-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/fonts-desc-02-t-manual-expected.png:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/fonts-desc-02-t-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/fonts-desc-05-t-manual-expected.png:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/fonts-desc-05-t-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/text-fonts-203-t-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/text-fonts-204-t-manual-expected.txt:
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::collectComplexTextRuns):
(WebCore::capitalized): Deleted.
(WebCore::shouldSynthesize): Deleted.
* Source/WebCore/platform/graphics/ComposedCharacterClusterTextIterator.h:
(WebCore::ComposedCharacterClusterTextIterator::reset):
* Source/WebCore/platform/graphics/Font.h:
(WebCore::Font::variantFont const):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::codePath const):
(WebCore::shouldSynthesizeSmallCaps):
(WebCore::capitalized):
* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebCore/platform/graphics/Latin1TextIterator.h:
(WebCore::Latin1TextIterator::Latin1TextIterator):
(WebCore::Latin1TextIterator::consume):
(WebCore::Latin1TextIterator::advance):
(WebCore::Latin1TextIterator::reset):
* Source/WebCore/platform/graphics/SurrogatePairAwareTextIterator.h:
(WebCore::SurrogatePairAwareTextIterator::SurrogatePairAwareTextIterator):
(WebCore::SurrogatePairAwareTextIterator::consume):
(WebCore::SurrogatePairAwareTextIterator::advance):
(WebCore::SurrogatePairAwareTextIterator::reset):
* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::hasExtraSpacing const):
(WebCore::resetGlyphBuffer):
(WebCore::SmallCapsState::SmallCapsState):
(WebCore::SmallCapsState::setSmallCapsData):
(WebCore::SmallCapsState::clear):
(WebCore::SmallCapsState::setIsSmallCaps):
(WebCore::SmallCapsState::skipSmallCapsProcessing const):
(WebCore::AdvanceInternalState::AdvanceInternalState):
(WebCore::AdvanceInternalState::fontChanged const):
(WebCore::AdvanceInternalState::updateFont):
(WebCore::WidthIterator::commitCurrentFontRange):
(WebCore::fontForRange):
(WebCore::WidthIterator::startNewFontRangeIfNeeded):
(WebCore::resetFontRangeIfNeeded):
(WebCore::updateCharacterAndSmallCapsIfNeeded):
(WebCore::WidthIterator::advanceInternal):
* Source/WebCore/platform/graphics/WidthIterator.h:

Canonical link: <a href="https://commits.webkit.org/267095@main">https://commits.webkit.org/267095@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f52a0c454605f580c14ec38ddb50a06d61234a86

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15638 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15944 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16316 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17394 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14673 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15820 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/18903 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16045 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17205 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15823 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16299 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13314 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18139 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13560 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14116 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21035 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14565 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14282 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17540 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14869 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12610 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14125 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3743 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18487 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14689 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -507,14 +507,9 @@ FontCascade::CodePath FontCascade::codePath(const TextRun& run, std::optional<un
-
-#else
 
+#if !(USE(FONT_VARIANT_VIA_FEATURES) || USE(FREETYPE))
```
Nit: I think this is easier to read if you distribute the `!`

litherum commented on Source/WebCore/platform/graphics/WidthIterator.h:
```diff
@@ -70,7 +70,7 @@ struct WidthIterator {
     };
     ApplyFontTransformsResult applyFontTransforms(GlyphBuffer&, unsigned lastGlyphCount, const Font&, CharactersTreatedAsSpace&);
-    void commitCurrentFontRange(GlyphBuffer&, unsigned& lastGlyphCount, unsigned currentCharacterIndex, const Font*&, const Font& newFont, const Font& primaryFont, UChar32 character, float& widthOfCurrentFontRange, float nextCharacterWidth, CharactersTreatedAsSpace&);
+    void commitCurrentFontRange(GlyphBuffer&, GlyphBuffer&, unsigned& lastGlyphCount, unsigned currentCharacterIndex, const Font*, const Font& primaryFont, float& widthOfCurrentFontRange, CharactersTreatedAsSpace&);
```
`GlyphBuffer&, GlyphBuffer&` could use some names to differentiate them

litherum commented on Source/WebCore/platform/graphics/WidthIterator.cpp:
```diff
@@ -196,21 +199,10 @@ void WidthIterator::commitCurrentFontRange(GlyphBuffer& glyphBuffer, unsigned& l
+        m_fallbackFonts->add(font);
 
     lastGlyphCount = glyphBuffer.size();
-    font = &newFont;
```
I think it's a shame that this line had to get deleted. I kind of feel that committing the current range should automatically set up as much state as possible, so the caller doesn't have to remember to do it themself.

litherum commented on Source/WebCore/platform/graphics/WidthIterator.cpp:
```diff
@@ -230,17 +222,35 @@ static void addToGlyphBuffer(GlyphBuffer& glyphBuffer, Glyph glyph, const Font&
+{
+    if (synthesizedFont)
+        return isSmallCaps ? smallSynthesizedFont : synthesizedFont;
+    return font;
```
How about

```
if (!synthesizedFont)
    return font;
return isSmallCaps ? smallSynthesizedFont : synthesizedFont;
```

litherum commented on Source/WebCore/platform/graphics/WidthIterator.cpp:
```diff
@@ -230,17 +222,35 @@ static void addToGlyphBuffer(GlyphBuffer& glyphBuffer, Glyph glyph, const Font&
     // The core logic here needs to match FontCascade::widthForSimpleText()
     bool rtl = m_run.rtl();
 
+    GlyphBuffer rangeBuffer;
```
Why did you opt to use a staging GlyphBuffer, and then copy its contents into the real glyph buffer during the commit step, rather than just using the original GlyphBuffer directly?

vitorroriz replied:
This was to avoid having to reset the glyphBuffer to the state of the last font transition. However, considering it again, I think it is not so difficult since we are already keeping track of the lastGlyphCount and just need to shrink the buffer to that size. I'll try fixing it.

litherum replied:
🆒

litherum commented on Source/WebCore/platform/graphics/WidthIterator.cpp:
```diff
@@ -637,13 +714,13 @@ void WidthIterator::advance(unsigned offset, GlyphBuffer& glyphBuffer)
 
     if (m_run.is8Bit()) {
-        Latin1TextIterator textIterator(m_run.data8(m_currentCharacterIndex), m_currentCharacterIndex, offset, length);
+        Latin1TextIterator textIterator(m_run.data8(0), m_currentCharacterIndex, offset, length);
```
Well this doesn't seem right.... WidthIterator's API is supposed to allow the caller to say 
```
myWidthIterator.advance(0);
myWidthIterator.advance(1);
myWidthIterator.advance(2);
```

litherum commented on Source/WebCore/platform/graphics/WidthIterator.cpp:
```diff
@@ -268,40 +297,86 @@ inline void WidthIterator::advanceInternal(TextIterator& textIterator, GlyphBuff
+        if (synthesizedFont || shouldSynthesizeCharacter)
+            glyph = nextRangeFont->glyphForCharacter(characterToWrite);
+        else
+            glyph = font->glyphForCharacter(characterToWrite);
```
this is a lot of code to embed in the middle of an already-big function that isn't going to be used most of the time. Is it possible to refactor this to move most of this logic to separate functions? Just to try to keep `advanceInternal()` somewhat readable?

litherum reviewed (commented):
I think this is generally the right direction but I'd like to hear what Vitor has to say about the staging GlyphBuffer rather than re-using the original GlyphBuffer before I go further.

litherum commented on Source/WebCore/platform/graphics/WidthIterator.cpp:
```diff
@@ -230,82 +206,240 @@ static void addToGlyphBuffer(GlyphBuffer& glyphBuffer, Glyph glyph, const Font&
         glyphBuffer.add(0, font, 0, currentCharacterIndex + 1);
 }
 
+struct SmallCapsState {
```
This is awesome

litherum commented on Source/WebCore/platform/graphics/WidthIterator.cpp:
```diff
@@ -230,82 +206,240 @@ static void addToGlyphBuffer(GlyphBuffer& glyphBuffer, Glyph glyph, const Font&
+    }
+};
+
+struct AdvanceState {
```
This is a bit confusing, because the word "advance" has 2 meanings:
- `advanceInternal()`
- each glyph has an advance (e.g `CTFontGetAdvancesForGlyphs()`)

This might be better if it was renamed. Perhaps `StateForAdvanceInternal`? `AdvanceInternalState`? Something like that.

litherum reviewed (approved):
I think this is a good approach.

vitorroriz committed:
Add support for font-variant-caps in WidthIterator (simple path)
https://bugs.webkit.org/show_bug.cgi?id=259736
rdar://110431790

Reviewed by Myles C. Maxfield.

Right now, font-variant small-caps can just be handled if we take
the Complex path in text layout, which is implemented by ComplexTextController.
Implementing small-caps handling in WidthIterator allows WebKit to
layout text taking the Simple path and use the WidthIterator when
font-feature or font-variants are used, including font-variant-caps.

The logic introduced for small-caps tries to mimic the one
in ComplexTextController. There, everytime there is change
in Font or in the small-caps state, a new run is created
with the determined font, which can be synthesized or not.
If any character in the run needs to be synthesized, all characters
are synthesized.

Here, for the WidthIterator, we don't split our runs in new runs for
each variation. Instead, we define a range and commit it. Each range
is represented by a font, that can be synthesized or not, like in
ComplexTextController, and a range glyph buffer. The buffer contains
the glyphs to the characters that compose the range. They are
add to the final glyph buffer each time we commit a range.

We are reusing the logic for determinig if a character should be synthesized
or not. Therefore, "shouldSynthesize()" and "capitalized()" functions
are moved from ComplexTextController to FontCascade.

When we are processing a range and we detect that the current font
can't be used for represent a glyph and it needs to be synthesized,
we need to synthesize the whole range. Therefore, we reset the iterator
to the index of the last font transition we have registered. For that,
we are adding the capability for the Latin1TextIterator and the
SurrogatePairAwareTextIterator to reset the current index to a given
index.

* LayoutTests/platform/gtk/fast/text/drawBidiText-expected.png:
* LayoutTests/platform/gtk/fast/text/drawBidiText-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/svg/import/text-dom-02-f-manual-expected.png: Added.
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/svg/import/text-dom-02-f-manual-expected.txt: Added.
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/svg/import/text-dom-05-f-manual-expected.png: Added.
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/svg/import/text-dom-05-f-manual-expected.txt: Added.
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/svg/import/text-tselect-02-f-manual-expected.png: Added.
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/svg/import/text-tselect-02-f-manual-expected.txt: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/svg/import/text-dom-02-f-manual-expected.png: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/svg/import/text-dom-02-f-manual-expected.txt: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/svg/import/text-dom-05-f-manual-expected.png: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/svg/import/text-dom-05-f-manual-expected.txt: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/svg/import/text-tselect-02-f-manual-expected.png: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/svg/import/text-tselect-02-f-manual-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/import/fonts-desc-02-t-manual.svg:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/text-dom-02-f-manual-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/text-dom-05-f-manual-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/text-tselect-02-f-manual-expected.txt:
* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/import/fonts-desc-02-t-manual-expected.png:
* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/import/fonts-desc-02-t-manual-expected.txt:
* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/import/fonts-desc-05-t-manual-expected.png:
* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/import/fonts-desc-05-t-manual-expected.txt:
* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/import/text-fonts-203-t-manual-expected.txt:
* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/import/text-fonts-204-t-manual-expected.txt:
* LayoutTests/platform/ios/svg/W3C-SVG-1.1/fonts-desc-02-t-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/fonts-desc-02-t-manual-expected.png:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/fonts-desc-02-t-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/fonts-desc-05-t-manual-expected.png:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/fonts-desc-05-t-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/text-fonts-203-t-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/text-fonts-204-t-manual-expected.txt:
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::collectComplexTextRuns):
(WebCore::capitalized): Deleted.
(WebCore::shouldSynthesize): Deleted.
* Source/WebCore/platform/graphics/ComposedCharacterClusterTextIterator.h:
(WebCore::ComposedCharacterClusterTextIterator::reset):
* Source/WebCore/platform/graphics/Font.h:
(WebCore::Font::variantFont const):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::codePath const):
(WebCore::shouldSynthesizeSmallCaps):
(WebCore::capitalized):
* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebCore/platform/graphics/Latin1TextIterator.h:
(WebCore::Latin1TextIterator::Latin1TextIterator):
(WebCore::Latin1TextIterator::consume):
(WebCore::Latin1TextIterator::advance):
(WebCore::Latin1TextIterator::reset):
* Source/WebCore/platform/graphics/SurrogatePairAwareTextIterator.h:
(WebCore::SurrogatePairAwareTextIterator::SurrogatePairAwareTextIterator):
(WebCore::SurrogatePairAwareTextIterator::consume):
(WebCore::SurrogatePairAwareTextIterator::advance):
(WebCore::SurrogatePairAwareTextIterator::reset):
* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::hasExtraSpacing const):
(WebCore::resetGlyphBuffer):
(WebCore::SmallCapsState::SmallCapsState):
(WebCore::SmallCapsState::setSmallCapsData):
(WebCore::SmallCapsState::clear):
(WebCore::SmallCapsState::setIsSmallCaps):
(WebCore::SmallCapsState::skipSmallCapsProcessing const):
(WebCore::AdvanceInternalState::AdvanceInternalState):
(WebCore::AdvanceInternalState::fontChanged const):
(WebCore::AdvanceInternalState::updateFont):
(WebCore::WidthIterator::commitCurrentFontRange):
(WebCore::fontForRange):
(WebCore::WidthIterator::startNewFontRangeIfNeeded):
(WebCore::resetFontRangeIfNeeded):
(WebCore::updateCharacterAndSmallCapsIfNeeded):
(WebCore::WidthIterator::advanceInternal):
* Source/WebCore/platform/graphics/WidthIterator.h:

Canonical link: https://commits.webkit.org/267095@main
